File: 8M3W/ehowlqk/boj_2239.js
const fs = require("fs");
const filePath =
  process.platform === "linux" ? "/dev/stdin" : "./8M3W/ehowlqk/boj_2239.txt";
const sudoku = fs
  .readFileSync(filePath)
  .toString()
  .split("\n")
  .map((line) => line.split("").map((num) => Number(num)));

const usedRow = Array.from({ length: 9 }, () => []);
const usedCol = Array.from({ length: 9 }, () => []);
const usedBlock = Array.from({ length: 3 }, () =>
  Array.from({ length: 3 }, () => [])
);

for (let i = 0; i < 9; i++) {
  for (let j = 0; j < 9; j++) {
    if (sudoku[i][j]) {
      usedRow[i].push(sudoku[i][j]);
      usedCol[j].push(sudoku[i][j]);
      usedBlock[Math.floor(i / 3)][Math.floor(j / 3)].push(sudoku[i][j]);
    }
  }
}

function backtracking(row, col) {
  if (row === 9) {
    for (let line of sudoku) {
      console.log(line.join(""));
    }
    return true;
  }

  const next = col === 8 ? [row + 1, 0] : [row, col + 1];

  if (sudoku[row][col]) return backtracking(...next);
  const candidates = check(row, col);
  for (let candidate of candidates) {
    sudoku[row][col] = candidate;
    usedRow[row].push(candidate);
    usedCol[col].push(candidate);
    usedBlock[Math.floor(row / 3)][Math.floor(col / 3)].push(candidate);
    if (backtracking(...next)) return true;
    sudoku[row][col] = 0;
    usedRow[row].pop();
    usedCol[col].pop();
    usedBlock[Math.floor(row / 3)][Math.floor(col / 3)].pop();
  }
}

function check(row, col) {
  const nums = Array.from({ length: 9 }, (_, i) => i + 1);
  return nums.filter(
    (e) =>
      !usedRow[row].includes(e) &&
      !usedCol[col].includes(e) &&
      !usedBlock[Math.floor(row / 3)][Math.floor(col / 3)].includes(e)
  );
}

backtracking(0, 0);

File: 8M3W/ehowlqk/boj_2239.py
def backtracking(row, col):
    # 마지막 depth에 도달했다면 정답 출력
    if row == 9:
        for line in sudoku:
            print(line, sep='')
        return True

    next = [row, col + 1] if col < 8 else [row + 1, 0]
    if sudoku[row][col] == 0:
        # 열, 행, 블록 체크 -> 가능한 숫자 목록 뽑기
        candidates = check(row, col)
        for candidate in candidates:
            sudoku[row][col] = candidate
            check_row[row].append(candidate)
            check_col[col].append(candidate)
            check_block[row//3][col//3].append(candidate)
            if backtracking(next[0], next[1]):
                return True
            sudoku[row][col] = 0
            check_row[row].pop()
            check_col[col].pop()
            check_block[row//3][col//3].pop()
    else:
        return backtracking(next[0], next[1])


def check(row, col):
    filter = check_row[row] + check_col[col] + check_block[row//3][col//3]
    return [x for x in range(1, 10) if x not in set(filter)]


sudoku, check_row, check_col, check_block = [], [[] for _ in range(9)], [[] for _ in range(9)], [[[] for _ in range(3)] for _ in range(3)]

for i in range(9):
    line = list(map(int, input()))
    sudoku.append(line)
    for j in range(9):
        # 각 행, 열, 블록 별 사용된 숫자 기록
        if line[j] > 0:
            check_row[i].append(line[j])
            check_col[j].append(line[j])
            check_block[i//3][j//3].append(line[j])

backtracking(0, 0)
